eth/catalyst/api.go

```diff
@@ -171,64 +174,92 @@ func (api *ConsensusAPI) makeEnv(parent *types.Block, header *types.Header) (*bl
+		return nil, &InvalidPayloadID
 	}
-	id := len(api.preparedBlocks)
-	api.preparedBlocks[id] = data
-	return &PayloadResponse{PayloadID: uint64(id)}, nil
-}
-
-func (api *ConsensusAPI) GetPayload(PayloadID hexutil.Uint64) (*ExecutableData, error) {
-	data, ok := api.preparedBlocks[int(PayloadID)]
+	id := binary.BigEndian.Uint64(hash[:8])
+	data, ok := api.preparedBlocks[id]
 	if !ok {
 		return nil, &UnknownPayload
 	}
 	return data, nil
 }
 
-// ConsensusValidated is called to mark a block as valid, so
-// that data that is no longer needed can be removed.
-func (api *ConsensusAPI) ConsensusValidated(params ConsensusValidatedParams) error {
-	switch params.Status {
-	case VALID.Status:
-		return nil
-	case INVALID.Status:
-		// TODO (MariusVanDerWijden) delete the block from the bc
-		return nil
-	default:
-		return errors.New("invalid params.status")
+func (api *ConsensusAPI) ForkchoiceUpdatedV1(heads ForkchoiceStateV1, PayloadAttributes *PayloadAttributesV1) (ForkChoiceResponse, error) {
+	if heads.HeadBlockHash == (common.Hash{}) {
+		return ForkChoiceResponse{Status: SUCCESS.Status, PayloadID: nil}, nil
 	}
+	if err := api.checkTerminalTotalDifficulty(heads.HeadBlockHash); err != nil {
+		if block := api.eth.BlockChain().GetBlockByHash(heads.HeadBlockHash); block == nil {
+			// TODO (MariusVanDerWijden) trigger sync
+			return SYNCING, nil
+		}
+		return INVALID, err
+	}
+	// If the finalized block is set, check if it is in our blockchain
+	if heads.FinalizedBlockHash != (common.Hash{}) {
+		if block := api.eth.BlockChain().GetBlockByHash(heads.FinalizedBlockHash); block == nil {
+			// TODO (MariusVanDerWijden) trigger sync
+			return SYNCING, nil
+		}
+	}
+	// SetHead
+	if err := api.setHead(heads.HeadBlockHash); err != nil {
+		return INVALID, err
+	}
+	// Assemble block (if needed)
+	if PayloadAttributes != nil {
+		data, err := api.assembleBlock(heads.HeadBlockHash, PayloadAttributes)
+		if err != nil {
+			return INVALID, err
+		}
+		hash := computePayloadId(heads.HeadBlockHash, PayloadAttributes)
+		id := binary.BigEndian.Uint64(hash)
+		api.preparedBlocks[id] = data
+		log.Info("Created payload", "payloadid", id)
+		// TODO (MariusVanDerWijden) do something with the payloadID?
+		hex := hexutil.Bytes(hash)
+		return ForkChoiceResponse{Status: SUCCESS.Status, PayloadID: &hex}, nil
+	}
+	return ForkChoiceResponse{Status: SUCCESS.Status, PayloadID: nil}, nil
 }
 
-func (api *ConsensusAPI) ForkchoiceUpdated(params ForkChoiceParams) error {
-	var emptyHash = common.Hash{}
-	if !bytes.Equal(params.HeadBlockHash[:], emptyHash[:]) {
-		if err := api.checkTerminalTotalDifficulty(params.HeadBlockHash); err != nil {
-			return err
-		}
-		return api.setHead(params.HeadBlockHash)
+func computePayloadId(headBlockHash common.Hash, params *PayloadAttributesV1) []byte {
+	time := make([]byte, 8)
+	binary.BigEndian.PutUint64(time, params.Timestamp)
+	// Hash
+	hasher := sha256.New()
+	hasher.Write(headBlockHash[:])
+	hasher.Write(time)
```

Comment:
I think this is equivalent to 
```
binary.Write(hasher, binary.BigEndian, uint64(params.Timestamp))
```
The `uint64` conversion is not needed there, but unless the linter rejects it, I think it might be nice to have there for clarity